changelogs/fragments/83292.yml

```diff
@@ -0,0 +1,3 @@
+---
+bugfixes:
+  - Fix the issue of playbook with any_errors_fatal as true, the rescue section only executes on the host where a fail task with run_once as true is triggered, not affecting other hosts as expected.(https://github.com/ansible/ansible/issues/83292)
```

Comment:
Can you add a test case without `any_errors_fatal: true` too? Since `run_once` is equivalent to `any_errors_fatal`, I'd expect the same result regardless. I think we need a similar fix for these two spots https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/linear.py#L203-L204, https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/linear.py#L216-L217 which also conflate the two features.

@mkrizek would know better, but maybe something like...
```diff
diff --git a/lib/ansible/plugins/strategy/linear.py b/lib/ansible/plugins/strategy/linear.py
index 29f94c4699..873adc9bd9 100644
--- a/lib/ansible/plugins/strategy/linear.py
+++ b/lib/ansible/plugins/strategy/linear.py
@@ -200,7 +200,7 @@ class StrategyModule(StrategyBase):
                         results.extend(self._execute_meta(task, play_context, iterator, host))
                         if task.args.get('_raw_params', None) not in ('noop', 'reset_connection', 'end_host', 'role_complete', 'flush_handlers'):
                             run_once = True
-                        if (task.any_errors_fatal or run_once) and not task.ignore_errors:
+                        if (task.any_errors_fatal and not run_once) and not task.ignore_errors:
                             any_errors_fatal = True
                     else:
                         # handle step if needed, skip meta actions as they are used internally
@@ -213,7 +213,7 @@ class StrategyModule(StrategyBase):
 
                         run_once = templar.template(task.run_once) or action and getattr(action, 'BYPASS_HOST_LOOP', False)
 
-                        if (task.any_errors_fatal or run_once) and not task.ignore_errors:
+                        if (task.any_errors_fatal and not run_once) and not task.ignore_errors:
                             any_errors_fatal = True
 
                         if not callback_sent:
```

Reply:
Hi there, @s-hertel thanks for the review. I tried without the without the any_errors_fatal:true and you suggestion is spot on with changing https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/linear.py#L203-L204, https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/linear.py#L216-L217. Without these changes it would fail. 
![image](https://github.com/ansible/ansible/assets/69774548/731e595b-aac7-4d0f-8b9b-844430f48889)

Reply:
@s-hertel I had thorough look it seems like with your suggested change, it does fix some issues, however when you have both any_errors_fatal and run_once, it will still have some problem(the inital case would fail as the SHOULD NOT HAPPEN case will trigger). I think there are deeper issues with how this is implemented. I will close this PR since I dont think there is much I can do.